```
[nix-shell:~/.cache/nixpkgs-review/pr-310218]$ nix-shell -I nixpkgs=$PWD/nixpkgs -p 'pkgs.python3.withPackages (p: with p; [ skein ])'
this derivation will be built:
  /nix/store/xav2y68ays9p0hqlsx1xximdl0hgiqqg-python3-3.11.9-env.drv
building '/nix/store/xav2y68ays9p0hqlsx1xximdl0hgiqqg-python3-3.11.9-env.drv'...
created 258 symlinks in user environment

[nix-shell:~/.cache/nixpkgs-review/pr-310218]$ python
Python 3.11.9 (main, Apr  2 2024, 08:25:04) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import skein
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/__init__.py", line 1, in <module>
    from . import kv
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/kv.py", line 12, in <module>
    import grpc as _grpc
  File "/nix/store/2qkc40wsir4qrynfwfmqlbssbcvaincq-python3.12-grpcio-1.62.2/lib/python3.12/site-packages/grpc/__init__.py", line 22, in <module>
    from grpc import _compression
  File "/nix/store/2qkc40wsir4qrynfwfmqlbssbcvaincq-python3.12-grpcio-1.62.2/lib/python3.12/site-packages/grpc/_compression.py", line 20, in <module>
    from grpc._cython import cygrpc
ImportError: cannot import name 'cygrpc' from 'grpc._cython' (/nix/store/2qkc40wsir4qrynfwfmqlbssbcvaincq-python3.12-grpcio-1.62.2/lib/python3.12/site-packages/grpc/_cython/__init__.py)
>>>

[nix-shell:~/.cache/nixpkgs-review/pr-310218]$
exit

[nix-shell:~/.cache/nixpkgs-review/pr-310218]$ nix-shell -I nixpkgs=$PWD/nixpkgs -p 'pkgs.python312.withPackages (p: with p; [ skein ])'

[nix-shell:~/.cache/nixpkgs-review/pr-310218]$ python
Python 3.12.3 (main, Apr  9 2024, 08:09:14) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import skein
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/__init__.py", line 1, in <module>
    from . import kv
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/kv.py", line 17, in <module>
    from .model import (
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/model.py", line 13, in <module>
    from .objects import Enum, ProtobufMessage, Specification, required
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/objects.py", line 7, in <module>
    from .utils import format_list, ensure_unicode, datetime_to_millis
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/utils.py", line 9, in <module>
    from setuptools._distutils.version import LooseVersion
ModuleNotFoundError: No module named 'setuptools'
```

Build succeeds, but fails to import. I'm not sure how it's passing the `pythonImportsCheck = [ "skein" ];`. Looking into it.

--- a/pkgs/development/python-modules/skein/default.nix
+++ b/pkgs/development/python-modules/skein/default.nix
@@ -64,12 +64,12 @@ buildPythonPackage rec {
     "test_cli"
   ];
 
-  meta = with lib; {
+  meta = {
     homepage = "https://jcristharif.com/skein";
     description = "A tool and library for easily deploying applications on Apache YARN";
     mainProgram = "skein";
-    license = licenses.bsd3;
-    maintainers = with maintainers; [ alexbiehl illustris ];
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ alexbiehl illustris ];
     # https://github.com/NixOS/nixpkgs/issues/48663#issuecomment-1083031627
     # replace with https://github.com/NixOS/nixpkgs/pull/140325 once it is merged
     broken = lib.traceIf isPy27 "${pname} not supported on ${python.executable}" isPy27;

--- a/pkgs/development/python-modules/skein/default.nix
+++ b/pkgs/development/python-modules/skein/default.nix
@@ -2,6 +2,7 @@
 , callPackage
 , fetchPypi
 , isPy27
+, pythonOlder
 , lib
 , cryptography
 , grpcio
@@ -10,6 +11,8 @@
 , hadoop
 , pytestCheckHook
 , python
+, setuptools
+, versioneer
 }:
 
 buildPythonPackage rec {
@@ -25,7 +28,8 @@ buildPythonPackage rec {
   jarHash = "sha256-x2KH6tnoG7sogtjrJvUaxy0PCEA8q/zneuI969oBOKo=";
   skeinJar = callPackage ./skeinjar.nix { inherit pname version jarHash; };
 
-  propagatedBuildInputs = [ cryptography grpcio pyyaml ];
+  propagatedBuildInputs = [ cryptography grpcio pyyaml ]
+    ++ lib.optionals (!pythonOlder "3.12") [ setuptools ];
   buildInputs = [ grpcio-tools ];
 
   preBuild = ''
@@ -37,7 +41,16 @@ buildPythonPackage rec {
   postPatch = ''
     substituteInPlace skein/core.py --replace "'yarn'" "'${hadoop}/bin/yarn'" \
       --replace "else 'java'" "else '${hadoop.jdk}/bin/java'"
+    # Remove vendorized versioneer
+    rm versioneer.py
+  '' + lib.optionalString (!pythonOlder "3.12") ''
+    substituteInPlace skein/utils.py \
+      --replace-fail "distutils" "setuptools._distutils"
   '';
+
+  build-system = [
+    versioneer
+  ];
 
   pythonImportsCheck = [ "skein" ];
 

